pkg/checks/checks.go

```diff
@@ -168,3 +170,45 @@ func WaitForInfraServices(ctx context.Context, appclient *appclient.AppsV1Client
 
 	return nil
 }
+
+func CheckDNS(ctx context.Context, rc *resources.Client, cs *acsapi.ContainerService) error {
+
+	//check master FQDN from ipAddress resource
+	rs, err := rc.Get(ctx, cs.Properties.AzProfile.ResourceGroup, "Microsoft.Network", "", "publicIPAddresses", "ip-apiserver")
+	if err != nil {
+		return err
+	}
+	fqdn := rs.Properties.(map[string]interface{})["dnsSettings"].(map[string]interface{})["fqdn"]
+	if fqdn != cs.Properties.MasterProfile.FQDN {
+		return fmt.Errorf("ip-address fqdn %s do not match configuration %s", fqdn, cs.Properties.MasterProfile.FQDN)
+	}
+
+	//check router FQDN by quering k8s LoadBalancer and ipAddress associated with it
+	rs, err = rc.Get(ctx, cs.Properties.AzProfile.ResourceGroup, "Microsoft.Network", "", "loadBalancers", "kubernetes")
+	if err != nil {
+		return err
+	}
+
+	// iterate over k8s LoadBalancers and check if we have a match. Match should always exist, so failed by default.
+	// TODO: With multiple router this need to check multiple conditions
+	exist := false
+	for _, v := range rs.Properties.(map[string]interface{})["frontendIPConfigurations"].([]interface{}) {
```

Comment:
Instead of using generic Azure Go client its more efficient to use SDK of a specific provider: https://github.com/Azure/azure-sdk-for-go/blob/master/services/network/mgmt/2018-05-01/network/loadbalancerfrontendipconfigurations.go#L49.

This way you won't need to cast into map etc. and you a get typed object.

Reply:
Gave it a shot. Not sure if this makes this look better. Logic still stays the same, but now I need to handle 2 separate Azure clients (`eipc network.PublicIPAddressesClient, lbc network.LoadBalancerFrontendIPConfigurationsClient`). 

I prefre this casting solution over having 2 clients. As soon as you have more than 2 clients, `struct` is good to abstract. Next think both packages `healthcheck` and `check` needs access to it, this means I need to externalise it to separate package so I could avoid circular imports. I think if at any point we decide to externalise all clients to separate package this would make sense, now - dont know.

@kargakis @pweil-  what do you think? 

Example:
https://gist.github.com/mjudeikis/1123f67e262c0d2559e4cbb2e30e6720
Clients:
https://gist.github.com/mjudeikis/1123f67e262c0d2559e4cbb2e30e6720

Reply:
yeah, typed is the way to go.  Compile time safety wins out here :-)